admob/src/android/ad_request_converter.cc

```diff
@@ -35,70 +37,88 @@
                          ADREQUESTBUILDER_METHODS);
 
-AdRequestConverter::AdRequestConverter(AdRequest request) {
+jobject GetJavaAdRequestFromCPPAdRequest(const AdRequest& request,
+                                         admob::AdMobError* error) {
+  FIREBASE_ASSERT(error);
+  *error = kAdMobErrorNone;
+
   JNIEnv* env = ::firebase::admob::GetJNI();
   jobject builder = env->NewObject(
       ad_request_builder::GetClass(),
       ad_request_builder::GetMethodId(ad_request_builder::kConstructor));
 
-  // Keywords.
-  for (int i = 0; i < request.keyword_count; i++) {
-    jstring keyword_str = env->NewStringUTF(request.keywords[i]);
-    builder = util::ContinueBuilder(
-        env, builder,
-        env->CallObjectMethod(
-            builder,
-            ad_request_builder::GetMethodId(ad_request_builder::kAddKeyword),
-            keyword_str));
-    env->DeleteLocalRef(keyword_str);
-  }
-
   // Network Extras.
-  if (request.extras_count > 0) {
+  const std::map<std::string, std::map<std::string, std::string>>& extras =
+      request.extras();
+  for (auto adapter_iter = extras.begin(); adapter_iter != extras.end();
+       ++adapter_iter) {
+    std::string adapter_name = adapter_iter->first;
+
+    jclass adapter_class = util::FindClass(env, adapter_name.c_str());
+    if (adapter_class == nullptr) {
+      LogError(
+          "Failed to resolve extras class. Check that \"%s\""
+          " is present in your APK.",
+          adapter_name.c_str());
+      *error = kAdMobErrorAdNetworkClassLoadError;
+      env->DeleteLocalRef(builder);
+      return nullptr;
+    }
+
     jobject extras_bundle =
         env->NewObject(util::bundle::GetClass(),
                        util::bundle::GetMethodId(util::bundle::kConstructor));
 
-    for (int i = 0; i < request.extras_count; i++) {
-      jstring extra_key_str = env->NewStringUTF(request.extras[i].key);
-      jstring extra_value_str = env->NewStringUTF(request.extras[i].value);
+    for (auto extra_iter = adapter_iter->second.begin();
+         extra_iter != adapter_iter->second.end(); ++extra_iter) {
+      jstring extra_key_str = env->NewStringUTF(extra_iter->first.c_str());
+      jstring extra_value_str = env->NewStringUTF(extra_iter->second.c_str());
       env->CallVoidMethod(extras_bundle,
                           util::bundle::GetMethodId(util::bundle::kPutString),
                           extra_key_str, extra_value_str);
       env->DeleteLocalRef(extra_value_str);
       env->DeleteLocalRef(extra_key_str);
     }
 
-    jclass admob_adapter_class =
-        util::FindClass(env, admob::GetActivity(), kAdMobAdapterClassName);
-
-    // If the adapter wasn't found correctly, the app has a serious problem.
-    FIREBASE_ASSERT_MESSAGE(
-        admob_adapter_class,
-        "Failed to locate the AdMobAdapter class for extras. Check that "
-        "com.google.ads.mediation.admob.AdMobAdapter is present in your APK.");
```

Comment:
minor nit: Not sure our assert macro works this way, but can you re-use kAdMobAdapterClassName in this string? (if it's complicated at all, don't bother.)

Reply:
This code and the defaulting to kAdMobAdapterClassName was removed.  We now return errors if the user-supplied class names could not be resolved, and we log the name.   See `if (adapter_class == nullptr) {` above.

Reply:
Got it, thanks.